### PR TITLE
cli: add support for servers with self-signed certs (fix #4564)

### LINKED ISSUE #4564: cli: add a flag to skip tls verification
Add a new global flag called `--insecure-skip-tls-verify` with default value `false`. We'll then need to use a custom [http client with tls config ](https://pkg.go.dev/crypto/tls?tab=doc#Client) set to skip the verification.

This is required to make the CLI work with server endpoints that have self-signed certificates.

Refer to [kubectl](https://github.com/kubernetes/kubectl) code for details.

We should also add a flag to accept the CA certificate.

### PR DESCRIPTION
**Pull request by ShraddhaAg**
### Description

Added 2 flags, `--insecure-skip-tls-verification` to disable cert checking and `--certificate-authority` for self signed CA. 
### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR.

### Affected components

- [X] CLI
- [ ] Docs

### Related Issues

#4564

### CODE REVIEW

File: cli/cli.go [hunk 1]
```diff
@@ -169,6 +172,18 @@ func (s *ServerConfig) ParseEndpoint() error {
 	return nil
 }
 
+// GetServerClient returns the http client for server
+func (s *ServerConfig) GetServerClient(config *tls.Config) *http.Client {
+	if config != nil {
```
Comment (arvi3411301) [inline cli/cli.go:177]:
```suggestion
	client := &http.Client{Transport: http.DefaultTransport}
	if config != nil {
	    tr := &http.Transport{TLSClientConfig: config}
	    client.Transport = tr
	}
	return client
```

File: cli/cli.go [hunk 2]
```diff
@@ -344,6 +371,44 @@ func (ec *ExecutionContext) Prepare() error {
...
+		cert, err := ec.getCAData(ec.CAPath)
+		if err != nil {
+			return err
+		}
+		if ok := rootCAs.AppendCertsFromPEM(cert); !ok {
+			return errors.Errorf("Unable to append given CA cert.")
+		}
+		ec.TLSClientConfig = &tls.Config{
+			RootCAs:            rootCAs,
+			InsecureSkipVerify: true,
```
Comment (arvi3411301) [inline cli/cli.go:394]:
```suggestion
			InsecureSkipVerify: ec.InsecureSkipTLSVerify,
```

File: cli/cli.go [hunk 3]
```diff
@@ -344,6 +371,44 @@ func (ec *ExecutionContext) Prepare() error {
...
+		}
+	} else {
```
Comment (arvi3411301) [inline cli/cli.go:396]: we can remove this

File: CHANGELOG.md [hunk 4]
```diff
@@ -35,6 +35,7 @@ Read more about the session argument for computed fields in the [docs](https://h
 ### Bug fixes and improvements
 
 - console: make GraphiQL Explorer taking the whole viewport
+- cli: add insecure-skip-tls-verification and certificate-authority flags (close #4564) (#4582)
```
Comment (tirumaraiselvan) [inline CHANGELOG.md:38]: This is just a minor clean-up. Could you move all the cli entries below console entries?

Comment (shahidhk) [reply CHANGELOG.md:38]: I am adding a detailed changelog for this.

**Comment by shahidhk** [comment]
We are making some tweaks to use this only for requests made to graphql-engine server.

File: CHANGELOG.md [hunk 5]
```diff
@@ -32,10 +32,18 @@ Support for this is now added through the `add_computed_field` API.
 
 Read more about the session argument for computed fields in the [docs](https://hasura.io/docs/1.0/graphql/manual/api-reference/schema-metadata-api/computed-field.html).
 
+### CLI: Support servers with self-signed certificates (close #4564) (#4582)
+
+A new global flag is added to the CLI for working with GraphQL Engine servers with
```
Comment (tirumaraiselvan) [inline CHANGELOG.md:37]: Cool. 

Minor: the first line and second line are both similar. If you read it, it seems little repeating. Can we combine then to be one line?

File: cli/cli.go [hunk 6]
```diff
@@ -148,8 +151,16 @@ type ServerConfig struct {
 	AdminSecret string `yaml:"admin_secret,omitempty"`
 	// APIPaths (optional) API paths for server
 	APIPaths *ServerAPIPaths `yaml:"api_paths,omitempty"`
+	// InsecureSkipTLSVerify - indicates if TLS verification is disabled or not.
+	InsecureSkipTLSVerify bool `yaml:"insecure_skip_tls_verify,omitempty"`
+	// CAPath - Path to a cert file for the certificate authority
+	CAPath string `json:"certificate_authority,omitempty"`
```
Comment (shahidhk) [inline cli/cli.go:157]: Can we change this to `ca_cert_path`?

Comment (arvi3411301) [reply cli/cli.go:157]: @shahidhk what about the flag?

Comment (shahidhk) [reply cli/cli.go:157]: okay, let it be then

File: docs/graphql/manual/hasura-cli/reference.rst [hunk 7]
```diff
@@ -157,6 +167,16 @@ The configuration can also be set in the form of environment variables:
      - Access key for Hasura GraphQL engine. Note: Deprecated. Use admin 
        secret instead. 
 
+   * - ``HASURA_GRAPHQL_INSECURE_SKIP_TLS_VERIFY``
+     - ``insecure_skip_tls_verify``
+     - Skip verifying SSL certificate for Hasura Endpoint. Useful if you have
```
Comment (marionschleifer) [inline docs/graphql/manual/hasura-cli/reference.rst:172]:
```suggestion
     - Skip verifying SSL certificate for the Hasura endpoint. Useful if you have
```

File: docs/graphql/manual/hasura-cli/reference.rst [hunk 8]
```diff
@@ -157,6 +167,16 @@ The configuration can also be set in the form of environment variables:
...
+       a self-singed certificate and don't have access to the CA cert.
+
+   * - ``HASURA_GRAPHQL_CERTIFICATE_AUTHORITY``
+     - ``certificate_authority``
+     - Path to the CA certificate for validating the self-signed certificate for
+       Hasura Endpoint.
```
Comment (marionschleifer) [inline docs/graphql/manual/hasura-cli/reference.rst:178]:
```suggestion
       the Hasura endpoint.
```

**Comment by coco98** [comment]
@shahidhk @ShraddhaAg Folks, could you link me to the docs for this feature please?

**Comment by ShraddhaAg** [comment]
@coco98 we updated it in the [CLI Configuration Reference](https://hasura.io/docs/1.0/graphql/manual/hasura-cli/config-reference.html), with the config file keys, env vars available for enabling this.

**Comment by coco98** [comment]
@ShraddhaAg Thanks. I was searching for the CLI flags reference. What the name of the flag is when used with the CLI, and I couldn't find it?

**Comment by ShraddhaAg** [comment]
@coco98 the two flags are `--certificate-authority` and `--insecure-skip-tls-verification`. There's more information [here](https://github.com/hasura/graphql-engine/blob/master/CHANGELOG.md#cli-support-servers-with-self-signed-certificates-close-4564-4582.). Looks like the CLI docs needs to be updated, I'm on it! 
(This may have been due to the cli command to generate rst docs wasn't working which was recently [fixed](https://github.com/hasura/graphql-engine/pull/4698).)

**Pull request merged**